### PR TITLE
Verify authentication tag length 

### LINKED ISSUE #1641: Extra characters at the end of encrypted JWT tokens don't prevent them from being considered valid
Hey,

Yesterday, the following report was posted on the OpenIddict repository: https://github.com/openiddict/openiddict-core/issues/1254.

I was able to reproduce it on my machine and confirmed this was caused by IdentityModel.
It looks like IdentityModel doesn't reject a JWE payload if it contains extra characters after the authentication tag.

Here's a repro:

```csharp
using System.Diagnostics;
using System.Security.Claims;
using System.Security.Cryptography;
using Microsoft.IdentityModel.JsonWebTokens;
using Microsoft.IdentityModel.Tokens;

namespace ConsoleApp13
{
    class Program
    {
        static void Main(string[] args)
        {
            var identity = new ClaimsIdentity(TokenValidationParameters.DefaultAuthenticationType);
            var signingKey = new RsaSecurityKey(RSA.Create(2048))
            {
                KeyId = "id1"
            };

            var encryptionKey = new RsaSecurityKey(RSA.Create(2048))
            {
                KeyId = "id2"
            };

            var handler = new JsonWebTokenHandler();
            var token = handler.CreateToken(new SecurityTokenDescriptor
            {
                Audience = "audience",
                Issuer = "issuer",
                EncryptingCredentials = new EncryptingCredentials(encryptionKey, SecurityAlgorithms.RsaOAEP, SecurityAlgorithms.Aes256CbcHmacSha512),
                SigningCredentials = new SigningCredentials(signingKey, SecurityAlgorithms.RsaSha256Signature),
                Subject = identity
            });

            var parameters = new TokenValidationParameters
            {
                ValidAudience = "audience",
                ValidIssuer = "issuer",
                TokenDecryptionKey = encryptionKey,
                IssuerSigningKey = signingKey
            };

            var result = handler.ValidateToken(token + "A", parameters);
            Debug.Assert(!result.IsValid, "The token shouldn't be considered valid.");
        }
    }
}
```

@brentschmaltz @jennyf19 could you please take a look?

### LINKED ISSUE #2201: [Bug] JWE is still valid after appending data to the Authenticaton Tag
**Which version of Microsoft.IdentityModel are you using?**
System.IdentityModel.Tokens.Jwt 7.1.0-preview

**Where is the issue?**
 * [x] M.IM.JsonWebTokens
 * [ ] M.IM.KeyVaultExtensions
 * [ ] M.IM.Logging
 * [ ] M.IM.ManagedKeyVaultSecurityKey
 * [ ] M.IM.Protocols
 * [ ] M.IM.Protocols.OpenIdConnect
 * [ ] M.IM.Protocols.SignedHttpRequest
 * [ ] M.IM.Protocols.WsFederation
 * [ ] M.IM.TestExtensions
 * [x] M.IM.Tokens
 * [ ] M.IM.Tokens.Saml
 * [ ] M.IM.Validators
 * [ ] M.IM.Xml
 * [x] S.IM.Tokens.Jwt
 * Other (please describe)

**Is this a new or an existing app?**
This is a new app or an experiment.

**Repro**

```csharp
using System.Security.Cryptography;
using Microsoft.IdentityModel.JsonWebTokens;
using Microsoft.IdentityModel.Tokens;

var jwtHandler = new JsonWebTokenHandler();

var claims = new Dictionary<string, object>
{
    { "claim1", "value1" },
    { "claim2", "value2" },
};

var signingKey = RSA.Create();
var validationKey = RSA.Create(signingKey.ExportParameters(false));

var decryptionKey = RSA.Create();
var encryptionKey = RSA.Create(decryptionKey.ExportParameters(false));

var token = jwtHandler.CreateToken(new SecurityTokenDescriptor
{
    Claims = claims,
    SigningCredentials = new SigningCredentials(
        new RsaSecurityKey(signingKey),
        SecurityAlgorithms.RsaSha256),
    EncryptingCredentials = new EncryptingCredentials(
        new RsaSecurityKey(encryptionKey),
        SecurityAlgorithms.RsaOAEP,
        SecurityAlgorithms.Aes256CbcHmacSha512)
});

// altering the JWE by appending some stuff to the Authentication Tag
var invalidToken = token + "the_token_has_been_tampered_with";

var validationResult = jwtHandler.ValidateToken(invalidToken, new()
{
    IssuerSigningKey = new RsaSecurityKey(validationKey),
    TokenDecryptionKey = new RsaSecurityKey(decryptionKey),
    ValidateAudience = false,
    ValidateIssuer = false,
    ValidateLifetime = false
});

if (validationResult.IsValid) // token is valid when it shouldn't be
    throw new Exception("The invalid token has been incorrectly marked as valid!");
```

**Expected behavior**
The validation of the JWE should fail, as the Authentication Tag has been altered, thus no longer valid.

Extracted from [RFC 7516, Section 5.2](https://datatracker.ietf.org/doc/html/rfc7516#section-5.2), regarding JWE decryption:
```
2.   Base64url decode the encoded representations of the JWE
       Protected Header, the JWE Encrypted Key, the JWE Initialization
       Vector, the JWE Ciphertext, the JWE Authentication Tag, and the
       JWE AAD, following the restriction that no line breaks,
       whitespace, or other additional characters have been used.
```

**Actual behavior**
The JWE is successfully validated when the Authentication Tag has been altered.

**Possible solution**
Check the actual length of the Authentication Tag and fail the validation if it does not match the expected value, according to the JWE encryption algorithm used.

### PR DESCRIPTION
**Pull request by kllysng**
[Design linked in Bug](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/2201#issuecomment-2072883307)

Fixes #2201 , #1641

### CODE REVIEW

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs [hunk 1]
```diff
@@ -165,13 +180,19 @@ private AuthenticatedEncryptionResult EncryptWithAesCbc(byte[] plaintext, byte[]
...
+                && expectedTagLength != authenticationTag.Length)
+                throw LogHelper.LogExceptionMessage(new SecurityTokenDecryptionFailedException(LogHelper.FormatInvariant(LogMessages.IDX10625, Base64UrlEncoder.Encode(authenticationTag))));
+
             byte[] al = Utility.ConvertToBigEndian(authenticatedData.Length * 8);
             byte[] macBytes = new byte[authenticatedData.Length + iv.Length + ciphertext.Length + al.Length];
             Array.Copy(authenticatedData, 0, macBytes, 0, authenticatedData.Length);
             Array.Copy(iv, 0, macBytes, authenticatedData.Length, iv.Length);
             Array.Copy(ciphertext, 0, macBytes, authenticatedData.Length + iv.Length, ciphertext.Length);
             Array.Copy(al, 0, macBytes, authenticatedData.Length + iv.Length + ciphertext.Length, al.Length);
-            if (!_symmetricSignatureProvider.Value.Verify(macBytes, 0, macBytes.Length, authenticationTag, 0, _authenticatedkeys.Value.HmacKey.Key.Length, Algorithm))
```
Comment (kllysng) [inline src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: @brentschmaltz we discussed possibly replacing authenticatedkeys.Value.HmacKey.Key.Length like this. It would remove the necessity for the check on lines 183-188, but then it wouldn't be behind the app context switch. I propose that we keep it as is (_authenticatedkeys.Value.HmacKey.Key.Length) and update it upon the next major release since we'll remove the context switch at that point.

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs [hunk 2]
```diff
@@ -32,6 +33,20 @@ private struct AuthenticatedKeys
...
         private EncryptionDelegate EncryptFunction;
         private const string _className = "Microsoft.IdentityModel.Tokens.AuthenticatedEncryptionProvider";
+        internal const string _skipValidationOfAuthenticationTagLength = "Switch.Microsoft.IdentityModel.SkipAuthenticationTagLengthValidation"; 
+
+        /// <summary>
+        /// Mapping from algorithm to the expected authentication tag length.
+        /// </summary>
+        internal static readonly Dictionary<string, int> ExpectedAuthenticationTagBase64UrlSizeInBytes = new()
+        {
+            { SecurityAlgorithms.Aes128Gcm, 16 },
```
Comment (kllysng) [inline src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: @brentschmaltz while testing, I discovered that the auth length issue does not repro with the AesGcm algos - do we anticipate needing these values in the future, or shall I remove them?

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs [hunk 3]
```diff
@@ -32,6 +33,20 @@ private struct AuthenticatedKeys
         private DecryptionDelegate DecryptFunction;
         private EncryptionDelegate EncryptFunction;
         private const string _className = "Microsoft.IdentityModel.Tokens.AuthenticatedEncryptionProvider";
+        internal const string _skipValidationOfAuthenticationTagLength = "Switch.Microsoft.IdentityModel.SkipAuthenticationTagLengthValidation"; 
+
+        /// <summary>
+        /// Mapping from algorithm to the expected authentication tag length.
+        /// </summary>
+        internal static readonly Dictionary<string, int> ExpectedAuthenticationTagBase64UrlSizeInBytes = new()
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: The map in SymmetricSignatureProvider has a lot of overlap

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs [hunk 4]
```diff
@@ -4189,108 +4188,228 @@ public static TheoryData<CreateTokenTheoryData> IncludeSecurityTokenOnFailureTes
...
+            Assert.Equal(theoryData.IsValid, tokenValidationResult.IsValid);
         }
 
-        public CreateTokenTheoryData(string testId) : base(testId)
+        public static TheoryData<CreateTokenTheoryData> ValidateAuthenticationTagLengthTheoryData()
         {
+            var signingCredentials512 = new SigningCredentials(KeyingMaterial.RsaSecurityKey_2048, SecurityAlgorithms.RsaPKCS1, SecurityAlgorithms.Sha512);
+            return new TheoryData<CreateTokenTheoryData>()
+            {
+                new("A128CBC-HS256_IsNotValidByDefault")
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:4246]: GCM cases?

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs [hunk 5]
```diff
@@ -4189,108 +4188,228 @@ public static TheoryData<CreateTokenTheoryData> IncludeSecurityTokenOnFailureTes
...
         {
+            // arrange
+            AppContext.SetSwitch(AuthenticatedEncryptionProvider._skipValidationOfAuthenticationTagLength, theoryData.EnableAppContextSwitch);
+            var payload = new JObject()
+            {
+                { JwtRegisteredClaimNames.Email, "Bob@contoso.com" },
+                { JwtRegisteredClaimNames.GivenName, "Bob" },
+                { JwtRegisteredClaimNames.Iss, "http://Default.Issuer.com"},
+                { JwtRegisteredClaimNames.Aud, "http://Default.Audience.com" },
+                { JwtRegisteredClaimNames.Iat, EpochTime.GetIntDate(DateTime.Parse("2017-03-17T18:33:37.095Z")).ToString() },
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs]: it is confusing to have hardcoded dates, something like DateTimeNow makes more sense.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs (hunk 4 above)
Comment (kllysng) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:4246]: I was not able to repro the bug when testing with AesGcm cases (mentioned [here](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2569#discussion_r1581717507))

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs (hunk 3 above)
Comment (kllysng) [reply src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: When I was chatting with Brent, he expressed that we shouldn't conflate auth tag and signature. Using "ExpectedSgnatureSizeInBytes" might be confusing. I'm happy to use it an replace this one altogether, since we may remove the AesGcm values as they don't replicate this bug, if we're okay with it being slightly confusing.

Comment (kllysng) [reply src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: I went ahead and used "SymmetricSignatureSizeInBytes" since a customer also mentioned it in the issue.

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs (hunk 2 above)
Comment (kllysng) [reply src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: I went ahead and removed the mappings as there are similar ones in SymmetricSignatureProvider. See this [comment](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2569/files#r1583822363)

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs (hunk 1 above)
Comment (kllysng) [reply src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: I went ahead and kept it "_authenticatedkeys.Value.HmacKey.Key.Length" for the same reason why we want an app context switch. We can update this on the next major rel of Wilson

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs (hunk 5 above)
Comment (kllysng) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs]: Happy to make this change - updated to DateTime.Now

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs (hunk 4 above)
Comment (kllysng) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:4246]: Closing this now - please reopen if needed!

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs [hunk 6]
```diff
@@ -32,6 +33,7 @@ private struct AuthenticatedKeys
         private DecryptionDelegate DecryptFunction;
         private EncryptionDelegate EncryptFunction;
         private const string _className = "Microsoft.IdentityModel.Tokens.AuthenticatedEncryptionProvider";
+        internal const string _skipValidationOfAuthenticationTagLength = "Switch.Microsoft.IdentityModel.SkipAuthenticationTagLengthValidation";
```
Comment (JoshLozensky) [inline src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs:36]: Super nit: can we add "Name" or "Switch" to the end of this variable? If I look at just the variable name "skipValidationOfAuthenticationTagLength" I would expect it to be a bool that tells me whether to skip this or not rather than the name of a switch.

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs [hunk 7]
```diff
@@ -165,6 +167,11 @@ private AuthenticatedEncryptionResult EncryptWithAesCbc(byte[] plaintext, byte[]
         private byte[] DecryptWithAesCbc(byte[] ciphertext, byte[] authenticatedData, byte[] iv, byte[] authenticationTag)
         {
             // Verify authentication Tag
+            if (ShouldValidateAuthenticationTagLength()
+                && SymmetricSignatureProvider.ExpectedSignatureSizeInBytes.TryGetValue(Algorithm, out int expectedTagLength)
```
Comment (JoshLozensky) [inline src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs:171]: Should the other two checks be added to the ShouldValidateAuthenticationTagLength method since they are necessary for making that decision?

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs [hunk 8]
```diff
@@ -4189,108 +4188,247 @@ public static TheoryData<CreateTokenTheoryData> IncludeSecurityTokenOnFailureTes
...
 
-    public class CreateTokenTheoryData : TheoryDataBase
-    {
-        public CreateTokenTheoryData()
+        [Theory, MemberData(nameof(ValidateAuthenticationTagLengthTheoryData))]
+        public void ValidateTokenAsync_ModifiedAuthNTag(CreateTokenTheoryData theoryData)
         {
+            // arrange
+            AppContext.SetSwitch(AuthenticatedEncryptionProvider._skipValidationOfAuthenticationTagLength, theoryData.EnableAppContextSwitch);
+            var payload = new JObject()
```
Comment (JoshLozensky) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:4197]: Consider putting the payload, jsonWebTokenHandler, and signingCredentials into the theory data.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs [hunk 9]
```diff
@@ -4189,108 +4188,247 @@ public static TheoryData<CreateTokenTheoryData> IncludeSecurityTokenOnFailureTes
...
+            {
+                { JwtRegisteredClaimNames.Email, "Bob@contoso.com" },
+                { JwtRegisteredClaimNames.GivenName, "Bob" },
+                { JwtRegisteredClaimNames.Iss, "http://Default.Issuer.com"},
+                { JwtRegisteredClaimNames.Aud, "http://Default.Audience.com" },
+                { JwtRegisteredClaimNames.Iat, EpochTime.GetIntDate(DateTime.Now).ToString() },
+                { JwtRegisteredClaimNames.Nbf, EpochTime.GetIntDate(DateTime.Now).ToString() },
+                { JwtRegisteredClaimNames.Exp, EpochTime.GetIntDate(DateTime.Now.AddDays(1)).ToString() },
```
Comment (JoshLozensky) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:4205]: Even if the number is one could you put it into a variable so we can see what it represents, otherwise it's just a magic #... i.e. I can see we are setting the expiry in a day, but why 1 day and if we ever want to change it having a variable to change is easier than reading through the code to find it.

**Review by JoshLozensky — APPROVED** [review]

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs (hunk 6 above)
Comment (kllysng) [reply src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs:36]: For consistency with our other context switch, I might keep is as is. https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/2421/files

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs (hunk 9 above)
Comment (kllysng) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:4205]: Thanks for the idea

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs [hunk 10]
```diff
@@ -134,6 +134,7 @@ internal static class LogMessages
         // public const string IDX10622 = "IDX10622:";
         // public const string IDX10623 = "IDX10623:";
         // public const string IDX10624 = "IDX10624:";
+        public const string IDX10625 = "IDX10625: Failed to verify the length of the authenticationTag '{0}'.";
```
Comment (GeoK) [inline src/Microsoft.IdentityModel.Tokens/LogMessages.cs]: Consider adding an aka.ms link that would point to an article on the [wiki](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki) explaining why one would start seeing this exception, how to set the config switch and what are the risks.
Log message seems misleading too, IdentityModel can verify, but verification fails as actual tag length doesn't correspond to the expected one.

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs (hunk 7 above)
Comment (kllysng) [reply src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs:171]: I appreciate your comment! The purpose of `ShouldValidateAuthenticationTagLength()` is to check if we need to validate the authentication tag length, not validate it, and would prefer to keep single responsibility for this method 😄

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs (hunk 10 above)
Comment (GeoK) [reply src/Microsoft.IdentityModel.Tokens/LogMessages.cs]: Adding the algorithm and actual/expected tag length to the message can also be helpful.

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs (hunk 8 above)
Comment (kllysng) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:4197]: Appreciate the idea.. All the theory data cases share the same payload, handler, and signing creds and there would be more code duplication to pull them out into the theory datas. 😵 Please let me know if there's another reason, otherwise I will keep as is 🤔

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs (hunk 2 above)
Comment (GeoK) [reply src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: @kellyyangsong I can't see the comment anymore. Does this PR apply only to AesCbc?

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs [hunk 11]
```diff
@@ -201,7 +213,6 @@ internal SymmetricSignatureProvider CreateSymmetricSignatureProvider()
             if (!IsSupportedAlgorithm(Key, Algorithm))
                 throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX10668, LogHelper.MarkAsNonPII(_className), LogHelper.MarkAsNonPII(Algorithm), Key)));
 
-            ValidateKeySize(Key, Algorithm);
```
Comment (GeoK) [inline src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: Why was this line removed?

Comment (kllysng) [reply src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: This might have been from a bad merge and should not have been removed. Thanks for the catch.

File: src/Microsoft.IdentityModel.Tokens/LogMessages.cs (hunk 10 above)
Comment (kllysng) [reply src/Microsoft.IdentityModel.Tokens/LogMessages.cs]: Added aka.ms link with a wiki page, updated log message, added algorithm, an expected/actual tag lengths.

File: src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs (hunk 11 above)
Comment (kllysng) [reply src/Microsoft.IdentityModel.Tokens/Encryption/AuthenticatedEncryptionProvider.cs]: Updated

**Review by GeoK — APPROVED** [review]

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs (hunk 8 above)
Comment (JoshLozensky) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandlerTests.cs:4197]: Defining the vars in the TheoryData method similar to signingCredentials512 and then assigning the variables to the CreateTokenTheoryData members "Payload", "JsonWebTokenHandler", and "SigningCredentials" would make the test cleaner though you are correct that total lines would increase as there would be 3 more lines per theory data. Currently 13 of the 30 lines (43%) in the test are just defining these variables and don't seem to add value to reader comprehension.

Ultimately your choice, imho since theory data is meant to hold the setup while the test is meant to hold the logic it seems to make more conceptual sense to put the vars in the theory data, esp. since it already has members for them but it is just a suggestion.
